extension/parquet/column_writer.cpp

```diff
@@ -1828,6 +1828,142 @@ void ListColumnWriter::FinalizeWrite(ColumnWriterState &state_p) {
+class ArrayColumnWriter : public ColumnWriter {
+public:
+	ArrayColumnWriter(ParquetWriter &writer, idx_t schema_idx, vector<string> schema_path_p, idx_t max_repeat,
+	                  idx_t max_define, unique_ptr<ColumnWriter> child_writer_p, bool can_have_nulls)
+	    : ColumnWriter(writer, schema_idx, std::move(schema_path_p), max_repeat, max_define, can_have_nulls),
+	      child_writer(std::move(child_writer_p)) {
+	}
+	~ArrayColumnWriter() override = default;
+
+	unique_ptr<ColumnWriter> child_writer;
+
+public:
+	unique_ptr<ColumnWriterState> InitializeWriteState(duckdb_parquet::format::RowGroup &row_group) override;
+	bool HasAnalyze() override;
+	void Analyze(ColumnWriterState &state, ColumnWriterState *parent, Vector &vector, idx_t count) override;
+	void FinalizeAnalyze(ColumnWriterState &state) override;
+	void Prepare(ColumnWriterState &state, ColumnWriterState *parent, Vector &vector, idx_t count) override;
+
+	void BeginWrite(ColumnWriterState &state) override;
+	void Write(ColumnWriterState &state, Vector &vector, idx_t count) override;
+	void FinalizeWrite(ColumnWriterState &state) override;
+};
+
+class ArrayColumnWriterState : public ColumnWriterState {
+public:
+	ArrayColumnWriterState(duckdb_parquet::format::RowGroup &row_group, idx_t col_idx)
+	    : row_group(row_group), col_idx(col_idx) {
+	}
+	~ArrayColumnWriterState() override = default;
+
+	duckdb_parquet::format::RowGroup &row_group;
+	idx_t col_idx;
+	unique_ptr<ColumnWriterState> child_state;
+	idx_t parent_index = 0;
+};
+
+unique_ptr<ColumnWriterState> ArrayColumnWriter::InitializeWriteState(duckdb_parquet::format::RowGroup &row_group) {
+	auto result = make_uniq<ArrayColumnWriterState>(row_group, row_group.columns.size());
+	result->child_state = child_writer->InitializeWriteState(row_group);
+	return std::move(result);
+}
+
+bool ArrayColumnWriter::HasAnalyze() {
+	return child_writer->HasAnalyze();
+}
+void ArrayColumnWriter::Analyze(ColumnWriterState &state_p, ColumnWriterState *parent, Vector &vector, idx_t count) {
+	auto &state = state_p.Cast<ArrayColumnWriterState>();
+	auto &array_child = ArrayVector::GetEntry(vector);
+	auto array_size = ArrayType::GetSize(vector.GetType());
+	child_writer->Analyze(*state.child_state, &state_p, array_child, array_size * count);
+}
+
+void ArrayColumnWriter::FinalizeAnalyze(ColumnWriterState &state_p) {
+	auto &state = state_p.Cast<ArrayColumnWriterState>();
+	child_writer->FinalizeAnalyze(*state.child_state);
+}
+
+void ArrayColumnWriter::Prepare(ColumnWriterState &state_p, ColumnWriterState *parent, Vector &vector, idx_t count) {
+	auto &state = state_p.Cast<ArrayColumnWriterState>();
+
+	auto array_size = ArrayType::GetSize(vector.GetType());
+	auto &validity = FlatVector::Validity(vector);
+
+	// write definition levels and repeats
+	idx_t start = 0;
+	idx_t vcount = parent ? parent->definition_levels.size() - state.parent_index : count;
+	idx_t vector_index = 0;
+	for (idx_t i = start; i < vcount; i++) {
```

Comment:
This is rather complex code and it looks like it's mostly a copy-paste from the `ListColumnWriter::Prepare` - could we perhaps unify the code (or just cast arrays to lists for parquet writing? we have a `CastColumnReader` for reading, we could similarly consider creating a `CastColumnWriter` for writing)

Reply:
The issue with unifying with the list is that we 1. don't have a list_entry_t child vector and 2. need to write validity differently. I think casting would be a good idea but I wasn't sure where to perform the cast. Ill have a look at the `CastColumnReader` and see if I can do something similar for the write-side.

Reply:
We can't template virtual functions so we'd have to template the whole class

Reply:
How about I just subclass the ListColumnWriter and override Analyze, Prepare, Write?